pkg/network/go/bininspect/symbols.go

```diff
@@ -127,38 +127,50 @@ func getSymbolsUnified(f *elf.File, typ elf.SectionType, wantedSymbols common.St
 	// Pre-allocating a buffer to read the symbol string into.
 	// The size of the buffer is maxSymbolNameSize + 1, for null termination.
 	symbolNameBuf := make([]byte, maxSymbolNameSize+1)
-	// Pre allocating a buffer for reading the symbol entry in the string table.
-	allocatedBufferForSymbolNameRead := make([]byte, 4)
-	// Pre allocating a buffer for reading the rest of the symbol fields from the symbol section.
-	allocatedBufferForSymbolRead := make([]byte, symbolSize-len(allocatedBufferForSymbolNameRead))
+
+	symbolSizeUint64 := uint64(symbolSize)
+	chunkSize := symbolChunkSize * symbolSizeUint64
+	symbolsCache := make([]byte, chunkSize)
+
+	var locationInCache int64
 
 	// Iterating through the symbol table. We skip the first symbolSize bytes as they are zeros.
-	for readLocation := int64(symbolSize); uint64(readLocation) < symbolSection.Size; readLocation += int64(symbolSize) {
-		// Reading the symbol entry in the string table.
-		stringEntry, err := readSymbolEntryInStringTable(symbolSection.ReaderAt, f.ByteOrder, readLocation, allocatedBufferForSymbolNameRead)
-		if err != nil {
-			log.Debugf("failed reading symbol entry %s", err)
+	for readLocation := symbolSizeUint64; readLocation < symbolSection.Size; readLocation += symbolSizeUint64 {
+		// Should read again
+		if (readLocation-symbolSizeUint64)/symbolSizeUint64%symbolChunkSize == 0 {
+			_, err := symbolSection.ReaderAt.ReadAt(symbolsCache, int64(readLocation))
+			if err != nil && err != io.EOF {
+				log.Debugf("failed reading symbol entry %s", err)
+				return nil, err
+			}
+			locationInCache = 0
+		} else {
+			locationInCache += int64(symbolSize)
+		}
+
+		stringEntry := int(f.ByteOrder.Uint32(symbolsCache[locationInCache : locationInCache+4]))
+		if stringEntry == 0 {
+			// symbol without a name.
 			continue
 		}
 
 		// Trying to get string representation of symbol.
 		// If the symbol name's length is not in the boundaries [minSymbolNameSize, maxSymbolNameSize+1] then we fail,
 		// and continue to the next symbol.
-		symbolName, ok := getSymbolNameByEntry(f.Sections[symbolSection.Link].ReaderAt, stringEntry, minSymbolNameSize, symbolNameBuf)
-		if !ok {
+		symbolNameSize := getSymbolNameByEntry(f.Sections[symbolSection.Link].ReaderAt, stringEntry, minSymbolNameSize, symbolNameBuf)
+
+		if symbolNameSize <= 0 {
 			continue
 		}
 
 		// Checking the symbol is relevant for us.
-		if _, ok := wantedSymbols[symbolName]; !ok {
+		if _, ok := wantedSymbols[string(symbolNameBuf[:symbolNameSize])]; !ok {
 			continue
 		}
 
-		// Complete the symbol reading.
 		var symbol elf.Symbol
-		if err := fillSymbol(&symbol, symbolSection.ReaderAt, f.ByteOrder, symbolName, readLocation+4, allocatedBufferForSymbolRead, is64Bit); err != nil {
-			continue
-		}
+		// Complete the symbol reading.
+		fillSymbol(&symbol, f.ByteOrder, string(symbolNameBuf[:symbolNameSize]), symbolsCache[locationInCache+4:locationInCache+int64(symbolSize)-4], is64Bit)
```

Comment:
`symbolsCache[locationInCache+4:locationInCache+int64(symbolSize)-4]` is hard to understand for the naive reader, try to set it as a variable explaining the meaning or adding a comment